gitlab/resource_gitlab_project.go

```diff
@@ -254,3 +295,70 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
 	}
 	return nil
 }
+
+func expandSharedWithGroupsOptions(d []interface{}) []*gitlab.ShareWithGroupOptions {
+	shareWithGroupOptionsList := []*gitlab.ShareWithGroupOptions{}
+
+	for _, config := range d {
+		data := config.(map[string]interface{})
+
+		groupAccess := accessLevelID[data["group_access"].(string)]
+
+		shareWithGroupOptions := &gitlab.ShareWithGroupOptions{
+			GroupID:     gitlab.Int(data["group_id"].(int)),
+			GroupAccess: &groupAccess,
+			ExpiresAt:   gitlab.String(data["expires_at"].(string)),
+		}
+
+		shareWithGroupOptionsList = append(shareWithGroupOptionsList,
+			shareWithGroupOptions)
+	}
+
+	return shareWithGroupOptionsList
+}
+
+func flattenSharedWithGroupsOptions(project *gitlab.Project) []interface{} {
+	sharedWithGroups := project.SharedWithGroups
+	sharedWithGroupsList := []interface{}{}
+
+	for _, option := range sharedWithGroups {
+		values := map[string]interface{}{
+			"group_id":     option.GroupID,
+			"group_access": accessLevel[gitlab.AccessLevelValue(option.GroupAccessLevel)],
+			"group_name":   option.GroupName,
+		}
+
+		sharedWithGroupsList = append(sharedWithGroupsList, values)
+	}
+
+	return sharedWithGroupsList
+}
+
+func updateSharedWithGroups(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	// Get existing groups and unshare all of them
+	project, _, err := client.Projects.GetProject(d.Id())
```

Comment:
This will probably cause issues from time to time if users interact with the project while it is being updated.

Couldn't you implement the same logic as for group members, e.g. only process the delta and not touch the unchanged shares?

Reply:
The issue here is that Gitlab API doesn't return the expires_at parameter. So there's no way knowing if it has changed. Plus, the API doesn't allow to edit this parameter but only to add or delete. Doing this way make it certain Gitlab has the expected information.

I agree it could create issues but it would still be very rare and the user would just have to retry its request for it to pass.

Reply:
In that case, the same logic still works: just remove and re-add the ones that change instead of updating them and do not touch the others.

Reply:
Regarding the absence of the `expires_at` parameter, is that true? 
The _Get a member of a group or project_ documentation says it is returnerd: https://docs.gitlab.com/ee/api/members.html#get-a-member-of-a-group-or-project

Reply:
Yes, that's true. The project object returned by the API looks like this `shared_with_groups: [
{
group_id: 306,
group_name: "DevOps2",
group_access_level: 40
}
],`

You can see the absence of expires_at even-though it's actually defined.

Reply:
And shared groups are not exposed in project's members API call

Reply:
Ok then, this is a sad limitation in the Gitlab API that we should probably document in a comment here.
I've opened an issue to see if this can/will change in the future https://gitlab.com/gitlab-org/gitlab-ce/issues/46107

In the mean time, I guess the `expires_at` should be removed from the management by the terraform provider if it cannot be read as it means it won't show in plans but applies will potentially change out of band modifications without knowing.

Reply:
Ok thanks!

I'll follow your recommendation and remove `expires_at` for now. And I'll add a check for modifications instead of unsharing all groups.

- [ ] Remove expires_at
- [ ] Improve shared_with_groups updates

Reply:
`expires_at` should become available in future releases:
https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/21104
https://gitlab.com/gitlab-org/gitlab-ce/issues/46107

Reply:
Good to know! It looks like they plan to release it in 11.3 / September 22.